3scale/3scale-operator
# Threescale 7668

# What 
[THREESCALE-7668](https://issues.redhat.com/browse/THREESCALE-7668)
Addition of new application CR, that supports :
- Create
- Update (Name, application plan, description)
- Delete
- Suspend 

# Verification 
1. ~As this [PR](https://github.com/3scale/3scale-porta-go-client/pull/48) with new functionality hasnt been merged yet, check out [this branch](https://github.com/Patryk-Stefanski/3scale-porta-go-client/tree/THREEESCALE-7668) of porta client locally.~

2. ~Check out this 3scale-operator branch locally and add the following line of code to go.mod. (This will use the local version of the porta client setup in step1)~
```bash
replace github.com/3scale/3scale-porta-go-client v0.6.0 =>  <LOCATION_OF_LOCAL_PORTA_CLIENT_REPO>
```

4. Run the following commands to install the 3scale operator
```bash
go mod vendor 
go mod tidy 
make install
export NAMESPACE=3scale-test
oc new-project $NAMESPACE
make download
make run
```

5. For the operator to create routes we need to mock the s3 credentials secret, run the following command.
```bash
kubectl apply -f - <<EOF
---
kind: Secret
apiVersion: v1
metadata: 
  name: s3-credentials
  namespace: 3scale-test
data: 
  AWS_ACCESS_KEY_ID: UkVQTEFDRV9NRQ==
  AWS_BUCKET: UkVQTEFDRV9NRQ==
  AWS_REGION: UkVQTEFDRV9NRQ==
  AWS_SECRET_ACCESS_KEY: UkVQTEFDRV9NRQ==
type: Opaque
EOF
```

6. Deploy the API manager CR  and specify your cluster wildcardDomain.
```bash
kubectl apply -f - <<EOF
---
apiVersion: apps.3scale.net/v1alpha1
kind: APIManager
metadata: 
  name: apimanager-sample
  namespace: 3scale-test
spec: 
  system: 
    fileStorage: 
      simpleStorageService: 
        configurationSecretRef: 
          name: s3-credentials
  wildcardDomain: apps.pstefans.****.s1.devshift.org
EOF
```

7. Create the required secrets for Accounts, products and backends. You can find the admin token in system-seed secret and the admin URL  under routes in the 3scale-test namespace.
```bash
kubectl apply -f - <<EOF
---
apiVersion: v1
kind: Secret
metadata:
  name: mytenant
type: Opaque
stringData:
  adminURL: <ADMIN_URL>
  token: "<ADMIN_ACCESS_TOKEN>"
EOF

kubectl apply -f - <<EOF
---
apiVersion: v1
kind: Secret
metadata:
  name: myusername01
stringData:
  password: "123456"
EOF
```

8. Create the developerAccount, DeveloperUser, product and backend required for the application CR.
```bash 
kubectl apply -f - <<EOF
---
apiVersion: capabilities.3scale.net/v1beta1
kind: DeveloperUser
metadata:
  name: developeruser01
  namespace: 3scale-test
spec:
  developerAccountRef:
    name: developeraccount01
  email: myusername01@example.com
  passwordCredentialsRef:
    name: myusername01
  providerAccountRef:
    name: mytenant
  role: admin
  username: myusername01
EOF

kubectl apply -f - <<EOF
---
apiVersion: capabilities.3scale.net/v1beta1
kind: DeveloperAccount
metadata:
  name: developeraccount01
  namespace: 3scale-test
spec:
  orgName: pstefans3
  providerAccountRef:
    name: mytenant
EOF


kubectl apply -f - <<EOF
---
apiVersion: capabilities.3scale.net/v1beta1
kind: Backend
metadata:
  name: backend1-cr
  namespace: 3scale-test
spec:
  mappingRules:
    - httpMethod: GET
      increment: 1
      last: true
      metricMethodRef: hits
      pattern: /
  name: backend1
  privateBaseURL: 'https://api.example.com'
  systemName: backend1
EOF

kubectl apply -f - <<EOF
---
apiVersion: capabilities.3scale.net/v1beta1
kind: Product
metadata:
  name: product1-cr
  namespace: 3scale-test
spec:
  applicationPlans:
    plan01:
      name: "My Plan 01"
      limits:
        - period: month
          value: 300
          metricMethodRef:
            systemName: hits
            backend: backend1
    plan02:
      name: "My Plan 02"
      limits:
        - period: month
          value: 300
          metricMethodRef:
            systemName: hits
            backend: backend1
  name: product1
  backendUsages:
    backend1:
      path: /
EOF
```

9.Create an Invalid Application CR , status should give out about invalid account and product CR names.
```bash
kubectl apply -f - <<EOF
---
apiVersion: capabilities.3scale.net/v1beta1
kind: Application
metadata:
  name: example
  namespace: 3scale-test
spec:
  accountCR: 
    name: developeraccount02
  applicationPlanName: plan03
  productCR: 
    name: product2-cr
  name: testApp
  description: further testing
EOF
```

10.  Update Application Resource , it should give out about the plan being used not belonging to the product.
```bash
kubectl apply -f - <<EOF
---
apiVersion: capabilities.3scale.net/v1beta1
kind: Application
metadata:
  name: example
  namespace: 3scale-test
spec:
  accountCR: 
    name: developeraccount01
  applicationPlanName: plan03
  productCR: 
    name: product1-cr
  name: testApp
  description: further testing
EOF
```

11. Update Application resource once again, this time the status should report ready: true with no error message. Check in 3scale UI that app has been created.
```bash
kubectl apply -f - <<EOF
---
apiVersion: capabilities.3scale.net/v1beta1
kind: Application
metadata:
  name: example
  namespace: 3scale-test
spec:
  accountCR: 
    name: developeraccount01
  applicationPlanName: plan01
  productCR: 
    name: product1-cr
  name: testApp
  description: further testing
EOF
```

11. Update the application one last time to check that updating the name , description , plan and suspending works.Check that the application update is reflected in 3scale UI and that the app has gone into the suspended state.
```bash
kubectl apply -f - <<EOF
---
apiVersion: capabilities.3scale.net/v1beta1
kind: Application
metadata:
  name: example
  namespace: 3scale-test
spec:
  accountCR: 
    name: developeraccount01
  applicationPlanName: plan02
  productCR: 
    name: product1-cr
  name: testApp12 
  description: further testing12
  suspend: true
EOF
```
